### DISCUSSION TITLE
Migrating monorepo to Bun.js: Dependency loop

### DISCUSSION THREAD

**Discussion by janwilmake** [oven-sh/bun#3350] [Q&A]
Hi there. I've been wanting to migrate my large typescript monorepo (300+ packages, 100.000+ lines of code) to bun quite a while now, and last few days I've finally given it a try! 

The first problem I encountered the one I'm still stuck at unfortunately: **Dependency loop**

<img width="574" alt="Screenshot 2023-06-18 at 07 36 38" src="https://github.com/oven-sh/bun/assets/97664551/ff5e536d-ef0e-47c5-861e-22773415d43c">

In the yarn workspace it installed fine, even with countless inter-package dependencies. The only problem sometimes was building. But to attempt to improve it, I spent some days building a system that detects which packages depend on each other and removed all the circular dependencies. Also I used [madge](https://github.com/pahen/madge) and cleaned up all inter-file circular dependencies. 

So I expected it to work after that.... but still! I get the same error.

As I couldn't find any issues about it, I wanted to post this here as anyone might have experience with it and solved it. Also maybe it would be nice to have some resources on how to smoothen the migration from large node.js projects. At the moment it's not really drop-in, imho.

My idea on what to do next is to sort all my packages into layers of dependency, and move them into bun one by one, until I find which one generates the dependency loop. But that'll have to wait until I can make another few days for this.

My theories of what caused it:

- Maybe bun is looking at the build folder instead of the src (where the typescript is)?
- Maybe one of my external dependencies has a "dependency loop"?

Maybe I'm completely looking at it in the wrong way, any thoughts/ideas are welcome!

**Comment by Jarred-Sumner** [comment]
We will change the error to show which package@version are conflicting with which dependency so that you can have more info on what's going wrong. Maybe something needs to be set as a peer dependency

If you're curious, this is the line where that happens:

https://github.com/oven-sh/bun/blob/83d7ec728f073261176b0bea8451b936c5a376ad/src/install/lockfile.zig#L475

**Reply by Jarred-Sumner** [reply]
![image](https://github.com/oven-sh/bun/assets/709451/41d7d03b-5f93-4db9-b4ae-737f037e6e4d)

https://github.com/oven-sh/bun/pull/3368

**Comment by janwilmake** [comment]
Thanks Jared, this is great news! I'll give it another go soon.
